### ISSUE TITLE
Allow unsafeDupableInterleaveIO to inline

### ISSUE THREAD

**Issue by clyring** [haskell/core-libraries-committee#242]
Currently, we have a `NOINLINE` pragma on `unsafeDupableInterleaveIO`. For subtle `-fstate-hack` reasons, this is needed to ensure that a call to `unsafeInterleaveIO` really executes its deferred action at most once. But this makes nearly every call to `unsafeDupableInterleaveIO` less efficient, and not just the ones that need the extra safety of `unsafeInterleaveIO`. The same reasoning applies for the `ST` versions of these functions.

I therefore propose to:
* Remove the `NOINLINE` pragmas on `unsafeDupableInterleaveIO` and `unsafeDupableInterleaveST`.
* Replace the `INLINE` pragmas on `unsafeInterleaveIO` and `unsafeInterleaveST` with `NOINLINE` pragmas.

I believe that doing so
* does not affect the "executed at most once" guarantee `unsafeInterleaveIO` is meant to provide,
* has no effect on the performance characteristics of `unsafeInterleaveIO`, and
* makes nearly every other of `unsafeDupableInterleaveIO` more efficient.

Downside: Doing so risks breaking code that for some reason explicitly calls `unsafeDupableInterleaveIO` with a `noDuplicate#`-using argument. Such code is hopefully rare; using `unsafeInterleaveIO` instead is generally preferable. But there is an example of such code inside GHC:

<details>

```haskell
mkSplitUniqSupply :: Char -> IO UniqSupply
-- ...documentation elided...
mkSplitUniqSupply c
  = unsafeDupableInterleaveIO (IO mk_supply)

  where
     !tag = mkTag c

        -- Here comes THE MAGIC: see Note [How the unique supply works]
        -- This is one of the most hammered bits in the whole compiler
        -- See Note [Optimising the unique supply]
        -- NB: Use noDuplicate# for thread-safety.
     mk_supply s0 =
        case noDuplicate# s0 of { s1 ->
        case unIO genSym s1 of { (# s2, u #) ->
        -- deferred IO computations
        case unIO (unsafeDupableInterleaveIO (IO mk_supply)) s2 of { (# s3, x #) ->
        case unIO (unsafeDupableInterleaveIO (IO mk_supply)) s3 of { (# s4, y #) ->
        (# s4, MkSplitUniqSupply (tag .|. u) x y #)
        }}}}

{-
Note [Optimising the unique supply]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The inner loop of mkSplitUniqSupply is a function closure

     mk_supply s0 =
        case noDuplicate# s0 of { s1 ->
        case unIO genSym s1 of { (# s2, u #) ->
        case unIO (unsafeDupableInterleaveIO (IO mk_supply)) s2 of { (# s3, x #) ->
        case unIO (unsafeDupableInterleaveIO (IO mk_supply)) s3 of { (# s4, y #) ->
        (# s4, MkSplitUniqSupply (tag .|. u) x y #)
        }}}}

It's a classic example of an IO action that is captured and then called
repeatedly (see #18238 for some discussion). It mustn't allocate!  The test
perf/should_run/UniqLoop keeps track of this loop.  Watch it carefully.

We used to write it as:

     mk_supply :: IO UniqSupply
     mk_supply = unsafeInterleaveIO $
                 genSym      >>= \ u ->
                 mk_supply   >>= \ s1 ->
                 mk_supply   >>= \ s2 ->
                 return (MkSplitUniqSupply (tag .|. u) s1 s2)

and to rely on -fno-state-hack, full laziness and inlining to get the same
result. It was very brittle and required enabling -fno-state-hack globally. So
it has been rewritten using lower level constructs to explicitly state what we
want.
-}
```
</details>

...and it appears that it is exactly the two uses of `unsafeDupableInterleaveIO` in `mk_supply` that are responsible for the observed GHC allocations improvements in the draft merge request [!11922](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11922) implementing this proposal.

**Comment by clyring** [comment]
Additional downside: If a user writes `x <- unsafeDupableInterleaveIO action ; y <- unsafeDupableInterleaveIO action` it is possible that `x` and `y` will get CSE'd. This can be mitigated by making the `State#` token returned by `unsafeDupableInterleaveIO` distinct from the state token it is given by inserting an opaque no-op, but there is not yet an appropriate primitive for doing so.

**Comment by simonpj** [comment]
To understand the original post, it helps to know these definitions:
```
noDuplicate :: IO ()
noDuplicate = IO $ \s -> case noDuplicate# s of s' -> (# s', () #)

{-# INLINE unsafeInterleaveIO #-}
unsafeInterleaveIO :: IO a -> IO a
unsafeInterleaveIO m = unsafeDupableInterleaveIO (noDuplicate >> m)

{-# NOINLINE unsafeDupableInterleaveIO #-}
-- See Note [unsafeDupableInterleaveIO should not be inlined]
unsafeDupableInterleaveIO :: IO a -> IO a
unsafeDupableInterleaveIO (IO m)
  = IO ( \ s -> let
                   r = case m s of (# _, res #) -> res
                in
                (# s, r #))
```
So currently `unsafeInterleaveIO` is just an (inlined) wrapper around `unsafeDupableInterleaveIO` which does the magic.

The proposal suggests changing `unsafeDupableInterleaveIO` into an INLINE function.

**Comment by tomjaguarpaw** [comment]
I don't have anywhere near enough understanding of the details of `INLINE`ing or GHC's implementation of IO to be able to make a correct decision on this. Accordingly, I'm happy to vote based on consensus of GHC experts.

**Comment by simonpj** [comment]
@clyring I have carefully re-read the comments in GHC.IO.Unsafe around these `unsafeInterleave` functions, and `Note [unsafeDupableInterleaveIO should not be inlined]`, and I just don't understand them.  (They were written by Simon Marlow 11 yrs ago, in the commit that fixed #5943, but that ticket provides no further clarification.  I am not sure:

* What the original problem was
* How `noDuplicate` + NOINLINE fixes it

Because of that I also don't know whether your proposed change still solves whatever problem the current arrangement solves.

---------------------

Aha.  Guided by a big of `git blame`, I found #11760.   I **think** that perhaps the problem that `GHC.IO.Unsafe` is trying to slve is this.  Consider
```
unsafeDupableInterleaveIO :: IO a -> IO a
unsafeDupableInterleaveIO (IO m)
  = IO ( \ s -> let
                   r = case m s of (# _, res #) -> res
                in
                (# s, r #))
```
That `r` is a thunk.  Now suppose that the thunk `r` is passed to two separate Concurrent Haskell threads, T1 and T2, and that both evaluate `r`.  If T1 completely evaluates `r` before T2 looks at it, all is fine.  But suppose T2 starts to evaluate `r` before T1 has black-holed it.  (With lazy black-holing, there may be a delay between starting to evaluate `r` and black-holing or updating it.)   In that case, T2 will evaluate the thunk as well as T1.  If the thunk is pure, no problem: they will get the same answer.  But if `(m s)` has side effects (e.g. does `putMVar`) the side effect may be done twice.  Yikes.

I am still very hazy how `noDuplicate` and NOINLINE on `unsafeInterleaveIO` solves this.

It's embarrassing how badly documented this is.

-------------------------

Futher insight.  By grepping around (there are no documentation pointer to this code) I found that `noDuplicate#` is implemented in the RTS in `rts/PrimOps.cmm`, line 2713.  There is a helpful comment there that explains how `noDuplicate#` does a runtime check for the race condition I describe above.    That helps to describe what how the `noDuplicate#` works; but it still doesn't explain why we must carefully NOINLINE.

Once we have this figured out, let's write  `Note [unsafeInterleaveIO magic]` as an overview to lay it all out.

**Comment by clyring** [comment]
@simonpj I agree that our existing documentation on the whole `unsafeInterleaveIO` implementation is extremely lacking. In fact improving the bad documentation is the real motivation for the MR, per its description:

> While trying to learn more about our black-hole mechanism(s), I found the reasoning in `Note [unsafeDupableInterleaveIO should not be inlined]` impossible to follow. So I replaced it.

I believe the real motivating issue for the `NOINLINE` pragma is [#5859](https://gitlab.haskell.org/ghc/ghc/-/issues/5859#note_59739). I believe the ticket [#5943](https://gitlab.haskell.org/ghc/ghc/-/issues/5943) is a duplicate, but have not _conclusively_ verified this yet. Alas, it seems the only reasoning on 5943 that got written down is the note that neither of us can make sense of today. But [your diagnosis of #5859](https://gitlab.haskell.org/ghc/ghc/-/issues/5859#note_59739) from the time makes sense and could still happen today if both `unsafeInterleaveIO` and `unsafeDupableInterleaveIO` get inlined.

------

The "additional downside" I mention in [this comment](https://github.com/haskell/core-libraries-committee/issues/242#issuecomment-1898808545) is worth fixing. So I think from CLC's perspective this proposal should wait until a proposal introducing the appropriate no-op primitive is introduced and approved.

**Comment by simonpj** [comment]
> I believe the real motivating issue for the NOINLINE pragma is [#5859](https://gitlab.haskell.org/ghc/ghc/-/issues/5859#note_59739). I believe the ticket [#5943](https://gitlab.haskell.org/ghc/ghc/-/issues/5943) is a duplicate, but have not conclusively verified this yet.

Aha.  I looked at these.  With some `git blame` and other spelunking, I don't think they are the same issue, but I think #5943 is indeed the reason for the NOINLINE.  Here is why.

We have in System.IO
```
fixIO :: (a -> IO a) -> IO a
fixIO k = do
    m <- newEmptyMVar
    ans <- unsafeDupableInterleaveIO
             (readMVar m `catch` \BlockedIndefinitelyOnMVar ->
                                    throwIO FixIOException)
    result <- k ans
    putMVar m result
    return result
```
That `unsafeDupableInterleaveIO` used to be `unsafeInterleaveIO`, but it was changed
to `unsafeDupableInterleaveIO` by this commit:
```
commit 239418cf94dede0f116bb859d1bb95891235eb76
Author: David Feuer <david.feuer@gmail.com>
Date:   Wed May 3 09:57:00 2017 -0400

    Improve fixIO

    Use `unsafeDupableInterleaveIO` to avoid `noDuplicate` calls. Switch
    from `takeMVar` to `readMVar` as multiple entry with `takeMVar`
    would lock things up.
```
Now, if we inline `unsafeDupableInterleaveIO` we get something like
```
fixIO k s
  = case newEmptyMVar s of (# s1, m #) ->
    let ans = case (readMVar m `catch` ...) s1 of (# _, res #) -> res
    case k ans s1 of (# s2, result #) ->
    case putMVar m result s2 of (# s3, () #) ->
    (# s3, result #)
```
Now, if the strictness analyser works out that `ans is used strictly (for
example by `k`, after inlining `fixIO`), it'll evaluate that `readMVar`
before we ever get to the `putMVar`.

And **that** is, I think, what was reported in #5943, and what explains the
otherwise rather opaque comment in `Note [unsafeDupableInterleaveIO should not be inlined]`
in GHC.IO.Unsafe, which says this:
```
-- Note [unsafeDupableInterleaveIO should not be inlined]
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
--
-- We used to believe that INLINE on unsafeInterleaveIO was safe,
-- because the state from this IO thread is passed explicitly to the
-- interleaved IO, so it cannot be floated out and shared.
--
-- HOWEVER, if the compiler figures out that r is used strictly here,
-- then it will eliminate the thunk and the side effects in m will no
-- longer be shared in the way the programmer was probably expecting,
-- but can be performed many times.  In #5943, this broke our
-- definition of fixIO, which contains
--
--    ans <- unsafeInterleaveIO (takeMVar m)
--
-- after inlining, we lose the sharing of the takeMVar, so the second
-- time 'ans' was demanded we got a deadlock.  We could fix this with
-- a readMVar, but it seems wrong for unsafeInterleaveIO to sometimes
-- share and sometimes not (plus it probably breaks the noDuplicate).
-- So now, we do not inline unsafeDupableInterleaveIO.
```
(Incidentally, it was originallly `takeMVar`, but the above commit
changed it to `readMVar` to avoid problems if it was executed twice.)

So **the purpose of the NOINLINE in `unsafeDupableInterleaveIO` is
to avoid the strictness analyser being able to spot that `r` is strict**.
Remember:
```
unsafeDupableInterleaveIO :: IO a -> IO a
{-# NOINLINE unsafeDupableInterleaveIO #-}
unsafeDupableInterleaveIO (IO m)
  = IO ( \ s -> let
                   r = case m s of (# _, res #) -> res
                in
                (# s, r #))
```
End of new insight

---------------

Hmm.  I wonder if a better approach would be to use `lazy` like this:
```
unsafeDupableInterleaveIO :: IO a -> IO a
unsafeDupableInterleaveIO (IO m)
  = IO ( \ s -> let
                   r = case m s of (# _, res #) -> res
                in
                (# s, lazy r #))    -- NB (lazy r)
```

**Comment by AndreasPK** [comment]
Having spent some hours debugging issues around `lazy` (see e.g. https://gitlab.haskell.org/ghc/ghc/-/issues/19181) so I think we should have a very high bar to making changes in these areas.

**Comment by clyring** [comment]
@simonpj I don't think your new hypothesis about [ghc#5943](https://gitlab.haskell.org/ghc/ghc/-/issues/5943) is plausible, for a few reasons.

1. Preventing `unsafeDupableInterleaveIO` from inlining has nothing to do with the `fixIO` failure mechanism you are considering. So it's a live bug (TODO: make ghc ticket), and it's not hard to construct a program that triggers this today:

    ```haskell
    import Control.Monad.Fix
    import Control.Exception
    import GHC.Exts (noinline)

    main :: IO ()
    main = do
      r <- mfix $ \p -> let
        {-# NOINLINE q #-}
        q = noinline id p
        in pure (True : q)
      print $! case r of { _:v:_ -> v ; _ -> False }
    ```

    For any law-abiding `MonadFix` instance, such a call to `mfix` should be semantically equivalent to `pure (repeat True)`. (Specifically, this follows from the 'Purity' law.) But if we compile the whole program with optimizations and inline `fixIO`, the variable `q` looks used-strictly, so CorePrep evaluates it eagerly, and we get a `FixIOException` where we shouldn't. To fix this, we can:
    * Prevent `fixIO` from inlining and hope we never improve strictness analysis enough to see that `q` is used strictly even in the un-inlined form, or
    * Hide strictness from the continuation of `fixIO`, so that `q` cannot be eagerly evaluated. (We currently don't have primops for this. Ticket: [ghc#23847](https://gitlab.haskell.org/ghc/ghc/-/issues/23847))
    * (Adding `lazy` in `unsafeDupablePerformIO` would not help!)

    (It's very plausible that allowing `unsafeDupablePerformIO` to inline would makes this bug _easier to trigger_, though.)

2. If we look at the [original program from the bug report ghc#5943](https://gitlab.haskell.org/ghc/ghc/-/snippets/694#LC100), we can clearly see a single `IO ()` action `net` that is executed twice (breaking the assumptions of the State Hack) and with comments suggesting it failed _only the second time it was executed_. The [test case](https://gitlab.haskell.org/ghc/ghc/-/blob/0cda2b8b15cdbc44c45ffa36a37ed8c2fe8b8b9c/libraries/base/tests/T5943.hs) added by the relevant patch is similar in this respect. (And the `putStrLn` call in the latter almost certainly hides any strictness the continuation might have had in `sig`/`s` anyway.)
3. Likewise, `Note [unsafeDupableInterleaveIO should not be inlined]` which you quoted talks about side effects "no longer be[ing] shared" and not about side effects being performed too soon.

--------------

> Having spent some hours debugging issues around use of lazy and unsafe IO functions in the past (see e.g. https://gitlab.haskell.org/ghc/ghc/-/issues/19181) I think we should have a very high bar to making changes in these areas.

I sadly agree that we must take move with great caution in this area. The specification of these unsafe `IO` functions is not as clear as we would like, and Hyrum's law is in full effect.

**Comment by simonpj** [comment]
You are dead right about fixIO.  That's a great example. I have created #24401.

Two simple solutions
* Do not inline `fixIO`
* Wrap the result of `fixIO` in `lazy`

That in turn would mean that we *can* inline `usafeDupableInterleaveIO`, I think.
After all, `usafePerformIO` and `unsafeDupablePerformIO` are inlined, so it's odd
that `unsafeInterleaveIO` is not.

**Comment by Bodigrim** [comment]
> The "additional downside" I mention in [this comment](https://github.com/haskell/core-libraries-committee/issues/242#issuecomment-1898808545) is worth fixing. So I think from CLC's perspective this proposal should wait until a proposal introducing the appropriate no-op primitive is introduced and approved.

Let me mark it as "dormant" then and close in the meantime. Feel free to reopen once you land the missing primop.

**Comment by treeowl** [comment]
>     1. Preventing `unsafeDupableInterleaveIO` from inlining has nothing to do with the `fixIO` failure mechanism you are considering. So it's a live bug (TODO: make ghc ticket), and it's not hard to construct a program that triggers this today:
>        ```haskell
>        import Control.Monad.Fix
>        import Control.Exception
>        import GHC.Exts (noinline)
>        
>        main :: IO ()
>        main = do
>          r <- mfix $ \p -> let
>            {-# NOINLINE q #-}
>            q = noinline id p
>            in pure (True : q)
>          print $! case r of { _:v:_ -> v ; _ -> False }
>        ```

That's terrible! It seems to me that we need to make sure that either `unsafeDupablePerformIO` or `fixIO` needs to wrap its result in `lazy`, and the former option is better.

**Comment by mpickering** [comment]
Another problem with `unsafeInterleaveIO` being `NOINLINE` is that you have to allocate a function before it is called, which is then subsequently retained by the thunk. If it was all inlined then this function allocation could be avoided. At the moment it makes using unsafeInterleaveIO quite memory heavy in you are being very careful about residency.
